src/Microsoft.ML.TimeSeries/SrCnnEntireAnomalyDetector.cs

```diff
@@ -0,0 +1,846 @@
+
+                    var detres = score > threshold ? 1 : 0;
+
+                    results[i][0] = detres;
+                    results[i][1] = score;
+                    results[i][2] = _ifftMagList[i];
+                }
+            }
+
+            private double[] BackAdd(double[] data)
+            {
+                AllocateDoubleArray(ref _predictArray, _lookaheadWindowSize + 1);
+                int j = 0;
+                for (int i = data.Length - _lookaheadWindowSize - 2; i < data.Length - 1; ++i)
+                {
+                    _predictArray[j++] = data[i];
+                }
+                var predictedValue = PredictNext(_predictArray);
+                AllocateDoubleArray(ref _backAddArray, data.Length + _backAddWindowSize);
+                for (int i = 0; i < data.Length; ++i)
+                {
+                    _backAddArray[i] = data[i];
+                }
+                for (int i = 0; i < _backAddWindowSize; ++i)
+                {
+                    _backAddArray[data.Length + i] = predictedValue;
+                }
+                return _backAddArray;
+            }
+
+            private double PredictNext(double[] data)
+            {
+                var n = data.Length;
+                double slopeSum = 0.0f;
+                for (int i = 0; i < n - 1; ++i)
+                {
+                    slopeSum += (data[n - 1] - data[i]) / (n - 1 - i);
+                }
+                return (data[1] + slopeSum);
+            }
+
+            private double[] AverageFilter(double[] data, int n)
+            {
+                double cumsum = 0.0f;
+                int length = data.Length;
+
+                AllocateDoubleArray(ref _cumSumList, length);
+                AllocateDoubleArray(ref _cumSumShift, length);
+
+                for (int i = 0; i < length; ++i)
+                {
+                    cumsum += data[i];
+                    _cumSumList[i] = cumsum;
+                    _cumSumShift[i] = cumsum;
+                }
+                for (int i = n; i < length; ++i)
+                {
+                    _cumSumList[i] = (_cumSumList[i] - _cumSumShift[i - n]) / n;
+                }
+                for (int i = 1; i < n; ++i)
+                {
+                    _cumSumList[i] /= (i + 1);
+                }
+                return _cumSumList;
+            }
+
+            private double CalculateScore(double mag, double avgMag)
+            {
+                double safeDivisor = avgMag;
+                if (Math.Abs(safeDivisor) < _eps)
+                {
+                    safeDivisor = _eps;
+                }
+                return (Math.Abs(mag - avgMag) / safeDivisor);
+            }
+
+            private void GetExpectedValue(double[] values, double[][] results)
+            {
+                //Step 8: Calculate Expected Value
+                var exps = CalculateExpectedValueByFft(GetDeanomalyData(values, GetAnomalyIndex(results.Select(x => x[1]).ToArray())));
```

Comment:
>CalculateExpectedValueByFft [](start = 27, length = 27)

Can return void, values are stored in a field.

Reply:
fixed.

---
In reply to: [427865834](https://github.com/dotnet/machinelearning/pull/5135#discussion_r427865834) [](ancestors = 427865834)